In [ ]:
import warnings
warnings.filterwarnings("ignore")

Read in the Data:

In [ ]:
import pandas as pd
import numpy as np
import csv
filename = "LoanStats_securev1_2019Q1.csv"
df = pd.read_csv(filename,header=1)
df.shape

(115677, 150)

We can see there are 150 features. The next steps of the code will be feature selection, which will find the most important features that we should be focusing on. Our target variable is Loan Status. We only want to focus on loans that are fully paid or charged off. Therefore, we will remove any rows that have something otherwise. 

In [ ]:
df = df[df['loan_status'].isin(['Fully Paid', 'Charged Off'])]

We will drop the features with the following criteria:

*   More than 50% of the data is missing
*   The feature would not have been available at the time of the loan application ("issue_d").

We will also address the following:

*  convert strings to numerical values
* drop superfluous attributes 
* highly correlated predictors
* zero (or nearly zero) variance predictors


This null values function below will give us a visual of the features with missing data and % of total values.

In [ ]:
def null_values(df):
        mis_val = df.isnull().sum()
        mis_val_percent = 100 * df.isnull().sum() / len(df)
        mis_val_table = pd.concat([mis_val, mis_val_percent], axis=1)
        mis_val_table_ren_columns = mis_val_table.rename(
        columns = {0 : 'Missing Values', 1 : '% of Total Values'})
        mis_val_table_ren_columns = mis_val_table_ren_columns[
            mis_val_table_ren_columns.iloc[:,1] != 0].sort_values(
        '% of Total Values', ascending=False).round(1)
        print ("Dataframe has " + str(df.shape[1]) + " columns.\n"      
            "There are " + str(mis_val_table_ren_columns.shape[0]) +
              " columns that have missing values.")
        return mis_val_table_ren_columns

In [ ]:
null_values(df)

Dataframe has 150 columns.
There are 61 columns that have missing values.


Missing Values  % of Total Values
member_id                       37761              100.0
next_pymnt_d                    37761              100.0
desc                            37761              100.0
settlement_percentage           37239               98.6
settlement_amount               37239               98.6
...                               ...                ...
last_pymnt_d                      224                0.6
dti                                98                0.3
revol_util                         47                0.1
all_util                            9                0.0
avg_cur_bal                         3                0.0

[61 rows x 2 columns]

Next, we will drop all variables that have more than 50% missing data. 

In [ ]:
missing_frac = df.isnull().mean()
drop_list = sorted(missing_frac[missing_frac >= 0.50].index)
print(drop_list)

['annual_inc_joint', 'debt_settlement_flag_date', 'deferral_term', 'desc', 'dti_joint', 'hardship_amount', 'hardship_dpd', 'hardship_end_date', 'hardship_last_payment_amount', 'hardship_length', 'hardship_loan_status', 'hardship_payoff_balance_amount', 'hardship_reason', 'hardship_start_date', 'hardship_status', 'hardship_type', 'member_id', 'mths_since_last_delinq', 'mths_since_last_major_derog', 'mths_since_last_record', 'mths_since_recent_bc_dlq', 'mths_since_recent_revol_delinq', 'next_pymnt_d', 'orig_projected_additional_accrued_interest', 'payment_plan_start_date', 'revol_bal_joint', 'sec_app_chargeoff_within_12_mths', 'sec_app_collections_12_mths_ex_med', 'sec_app_earliest_cr_line', 'sec_app_fico_range_high', 'sec_app_fico_range_low', 'sec_app_inq_last_6mths', 'sec_app_mort_acc', 'sec_app_mths_since_last_major_derog', 'sec_app_num_rev_accts', 'sec_app_open_acc', 'sec_app_open_act_il', 'sec_app_revol_util', 'settlement_amount', 'settlement_date', 'settlement_percentage', 'settlem

The function below will remove the variables in the drop list above from the dataset. 

In [ ]:
def drop_cols(cols):
    df.drop(labels=cols, axis=1, inplace=True)

In [ ]:
drop_cols(drop_list)

In [ ]:
df.shape

(37761, 106)

Next, I will remove the features that were unavailable before lending a loan.

In [ ]:
print(sorted(df.columns))

['acc_now_delinq', 'acc_open_past_24mths', 'addr_state', 'all_util', 'annual_inc', 'application_type', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util', 'chargeoff_within_12_mths', 'collection_recovery_fee', 'collections_12_mths_ex_med', 'debt_settlement_flag', 'delinq_2yrs', 'delinq_amnt', 'dti', 'earliest_cr_line', 'emp_length', 'emp_title', 'fico_range_high', 'fico_range_low', 'funded_amnt', 'funded_amnt_inv', 'grade', 'hardship_flag', 'home_ownership', 'id', 'il_util', 'initial_list_status', 'inq_fi', 'inq_last_12m', 'inq_last_6mths', 'installment', 'int_rate', 'issue_d', 'last_credit_pull_d', 'last_fico_range_high', 'last_fico_range_low', 'last_pymnt_amnt', 'last_pymnt_d', 'loan_amnt', 'loan_status', 'max_bal_bc', 'mo_sin_old_il_acct', 'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl', 'mort_acc', 'mths_since_rcnt_il', 'mths_since_recent_bc', 'mths_since_recent_inq', 'num_accts_ever_120_pd', 'num_actv_bc_tl', 'num_actv_rev_tl', 'num_bc_sats', 'num_bc_tl', 'num_il_tl', 'n

In [ ]:
drop_list2 = ['acc_now_delinq', 'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util', 'chargeoff_within_12_mths', 'collection_recovery_fee', 'collections_12_mths_ex_med', 'debt_settlement_flag', 'delinq_2yrs', 'delinq_amnt', 'funded_amnt', 'funded_amnt_inv', 'hardship_flag', 'inq_last_6mths', 'last_credit_pull_d', 'last_fico_range_high', 'last_fico_range_low', 'last_pymnt_amnt', 'last_pymnt_d', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl', 'mths_since_recent_bc', 'mths_since_recent_inq', 'num_accts_ever_120_pd', 'num_actv_bc_tl', 'num_actv_rev_tl', 'num_bc_sats', 'num_bc_tl', 'num_il_tl', 'num_op_rev_tl', 'num_rev_accts', 'num_rev_tl_bal_gt_0', 'num_sats', 'num_tl_120dpd_2m', 'num_tl_30dpd', 'num_tl_90g_dpd_24m', 'num_tl_op_past_12m',  'out_prncp', 'out_prncp_inv', 'pct_tl_nvr_dlq', 'percent_bc_gt_75', 'pymnt_plan', 'recoveries', 'tax_liens', 'tot_coll_amt', 'tot_cur_bal', 'tot_hi_cred_lim', 'total_bal_ex_mort', 'total_bc_limit', 'total_il_high_credit_limit', 'total_pymnt', 'total_pymnt_inv', 'total_rec_int', 'total_rec_late_fee', 'total_rec_prncp', 'total_rev_hi_lim']
drop_cols(drop_list2)
len(df.columns)

49

In [ ]:
df.head()

id  loan_amnt  ... mort_acc pub_rec_bankruptcies
1   149089105     6500.0  ...      1.0                  0.0
5   149477700    15000.0  ...      1.0                  0.0
11  149459186     1000.0  ...      0.0                  0.0
14  149521312     3025.0  ...      3.0                  0.0
15  149521158     2000.0  ...      0.0                  0.0

[5 rows x 49 columns]

Now we are left with 49 features, I will screen each feature and decide if a feature should be dropped. After pre-processing each feature, I will do some overall statistical tests for all the features, their correlations, etc. The tasks involved include:

* Missing data and outliers
* Cleaning & Formatting:
* Transformation of features
* Choose the most relevant features in the data

In [ ]:
drop_cols(['id','grade','emp_title','title','zip_code','policy_code', 'url','open_acc_6m', 'open_act_il', 'open_il_12m','open_il_24m', 'mths_since_rcnt_il', 'total_bal_il', 'il_util','open_rv_12m','open_rv_24m','open_rv_24m','max_bal_bc','all_util','inq_fi','total_cu_tl','inq_last_12m'])

In [ ]:
# Fix the term feature
df['term'].sample(5)
df['term'] = df['term'].apply(lambda s: np.int8(s.split()[0]))

In [ ]:
# Fix the emp_length feature (convert to integer)
df['emp_length'].replace('10+ years', '10 years', inplace=True)
df['emp_length'].replace('< 1 year', '0 years', inplace=True)
df['emp_length'].value_counts(dropna=False).sort_index()
df.emp_length.map( lambda x: str(x).split()[0]).value_counts(dropna=True).sort_index()
df['emp_length'] = df.emp_length.map(lambda x: float(str(x).split()[0]))
df['emp_length'].sample(5)

66116     6.0
77747     6.0
19028     0.0
47149     5.0
42952    10.0
Name: emp_length, dtype: float64

In [ ]:
# Fix the int_rate feature
df['int_rate'].replace(regex=True, inplace=True, to_replace=r'%', value=r'')
df['int_rate'] = df['int_rate'].apply(pd.to_numeric, errors='coerce')

In [ ]:
# Fix the annual income variable, use log-transform since there is a large range of variation, we log-transform the values.
df['annual_inc'] = df['annual_inc'].apply(lambda x:np.log10(x+1))

In [ ]:
# fix earliest_cr_line to datetime variable
from datetime import datetime

df.earliest_cr_line = pd.to_datetime(df.earliest_cr_line)
dttoday = datetime.now().strftime('%Y-%m-%d')
df.earliest_cr_line = df.earliest_cr_line.apply(lambda x:(np.timedelta64((x - pd.Timestamp(dttoday)),'D').astype(int))/-365)
df.earliest_cr_line.shape

(37761,)

In [ ]:
# create new variable fico_score, which is the mean value of fico_low and fico_high
df['fico_score'] = (df['fico_range_low'] + df['fico_range_high'])/2.
drop_cols(['fico_range_high','fico_range_low'])

In [ ]:
# Change the bin width of pub_rec
df.pub_rec = df.pub_rec.map(lambda x: 3 if x >2.0 else x)

In [ ]:
# Fix revol_bal
df['revol_bal'] = df['revol_bal'].apply(lambda x:np.log10(x+1))

In [ ]:
# Fix revol_util
df['revol_util'] = df['revol_util'].replace(regex=True, inplace=False, to_replace=r'%', value=r'')

In [ ]:
df['revol_util'] = df['revol_util'].apply(pd.to_numeric, errors='coerce')

In [ ]:
df.head()

loan_amnt  term  int_rate  ...  mort_acc pub_rec_bankruptcies  fico_score
1      6500.0    36      8.81  ...       1.0                  0.0       687.0
5     15000.0    36     17.97  ...       1.0                  0.0       662.0
11     1000.0    36      7.56  ...       0.0                  0.0       767.0
14     3025.0    36     17.19  ...       3.0                  0.0       667.0
15     2000.0    36     16.40  ...       0.0                  0.0       682.0

[5 rows x 27 columns]

In [ ]:
# Change target variable to be a binary variable. Fully Paid = 0, Charged Off = 1
df['Charged_Off'] = df['loan_status'].apply(lambda s: np.float(s == 'Charged Off'))
drop_cols('loan_status')
df.shape

(37761, 27)

In [ ]:
df.head()

loan_amnt  term  int_rate  ...  pub_rec_bankruptcies fico_score  Charged_Off
1      6500.0    36      8.81  ...                   0.0      687.0          0.0
5     15000.0    36     17.97  ...                   0.0      662.0          0.0
11     1000.0    36      7.56  ...                   0.0      767.0          0.0
14     3025.0    36     17.19  ...                   0.0      667.0          0.0
15     2000.0    36     16.40  ...                   0.0      682.0          0.0

[5 rows x 27 columns]

In [ ]:
# drop columns that are not linearly correlated with the target variable. 
drop_cols(['installment', 'mo_sin_old_rev_tl_op','total_acc','pub_rec_bankruptcies'])
df.shape

(37761, 23)

In [ ]:
df.head()

loan_amnt  term  int_rate  ... mort_acc  fico_score Charged_Off
1      6500.0    36      8.81  ...      1.0       687.0         0.0
5     15000.0    36     17.97  ...      1.0       662.0         0.0
11     1000.0    36      7.56  ...      0.0       767.0         0.0
14     3025.0    36     17.19  ...      3.0       667.0         0.0
15     2000.0    36     16.40  ...      0.0       682.0         0.0

[5 rows x 23 columns]

In [ ]:
dummy_list =['sub_grade','home_ownership','verification_status','purpose','addr_state','initial_list_status','application_type']
df[dummy_list].isnull().any()
df = pd.get_dummies(df, columns=dummy_list, drop_first=True)

I will use the earlier funded loans to predict the later funded loans

In [ ]:
df['issue_d'].sample()

2871    Mar-2019
Name: issue_d, dtype: object

In [ ]:
df['issue_d'] = pd.to_datetime(df['issue_d'])
df['issue_d'].sample()

68189   2019-02-01
Name: issue_d, dtype: datetime64[ns]

I will split the train/test data into 80/20

In [ ]:
df_train = df.loc[df['issue_d']  < df['issue_d'].quantile(0.8)]
df_test =  df.loc[df['issue_d'] >= df['issue_d'].quantile(0.8)]

In [ ]:
df_train.drop('issue_d', axis=1, inplace=True)
df_test.drop('issue_d', axis=1, inplace=True)

I will now separate the predictor variables from the target variable

In [ ]:
X_train = df_train.drop(['Charged_Off'], axis=1)
y_train = df_train.loc[:, 'Charged_Off']

X_test = df_test.drop(['Charged_Off'], axis=1)
y_test = df_test['Charged_Off']

In [ ]:
X_train.head()

loan_amnt  term  ...  initial_list_status_w  application_type_Joint App
22658    35000.0    36  ...                      1                           0
22659     5000.0    36  ...                      1                           0
22662     4500.0    36  ...                      1                           0
23196    14000.0    36  ...                      1                           0
23410     8000.0    36  ...                      1                           0

[5 rows x 113 columns]

Impute the missing variables

In [ ]:
from sklearn.impute import SimpleImputer
imputer = SimpleImputer(missing_values=np.nan, strategy='median')

# Train on the training features
imputer.fit(X_train)

# Transform both training and testing data
X_train = pd.DataFrame(imputer.transform(X_train), columns=X_train.columns)
X_test = pd.DataFrame(imputer.transform(X_test), columns=X_test.columns)

In [ ]:
from sklearn.preprocessing import StandardScaler
# Create an imputer object with a median filling strategy
scaler = StandardScaler()

# Train on the training features
scaler.fit(X_train)

# Transform both training and testing data
X_train = pd.DataFrame(scaler.transform(X_train), columns=X_train.columns)
X_test = pd.DataFrame(scaler.transform(X_test), columns=X_test.columns)

In [ ]:
from sklearn.linear_model import SGDClassifier
from sklearn.pipeline import Pipeline

kfold = 3
random_state = 42
pipeline_sgdlr = Pipeline([('model', SGDClassifier(loss='log', max_iter=1000, tol=1e-3, random_state=random_state, warm_start=True))])

param_grid_sgdlr  = {
    'model__alpha': [10**-5, 10**-1, 10**2],
    'model__penalty': ['l1', 'l2']}

grid_sgdlr = GridSearchCV(estimator=pipeline_sgdlr, param_grid=param_grid_sgdlr, scoring='roc_auc', n_jobs=-1, pre_dispatch='2*n_jobs', cv=kfold, verbose=1, return_train_score=False)
grid_sgdlr.fit(X_train, y_train)

sgdlr_estimator = grid_sgdlr.best_estimator_
print('Best score: ', grid_sgdlr.best_score_)
print('Best parameters set: \n', grid_sgdlr.best_params_)

Fitting 3 folds for each of 6 candidates, totalling 18 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done  18 out of  18 | elapsed:    9.7s finished


Best score:  0.6891919428126817
Best parameters set: 
 {'model__alpha': 0.1, 'model__penalty': 'l2'}


In [ ]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(n_jobs=-1, random_state=random_state, max_features= 'sqrt' ,n_estimators=50) 
param_grid_rf = {
    'n_estimators': [50], # The number of randomized trees to build 
    'class_weight': [{0:1, 1:1}] #'model__class_weight': [{0:1, 1:1}, {0:1,1:2}, {0:1, 1:3}, {0:1, 1:4}]
   # 'model__max_features': range(2,7),
   # 'model__min_samples_leaf':range(2,6),
}
grid_rf = GridSearchCV(estimator=rf, param_grid=param_grid_rf, scoring='roc_auc',n_jobs=-1,pre_dispatch='2*n_jobs', cv=kfold, verbose=1, return_train_score=False)
grid_rf.fit(X_train, y_train)
rf_estimator = grid_rf.best_estimator_
print('Best score: ', grid_rf.best_score_)
print('Best parameters set: \n', grid_rf.best_params_)

Fitting 3 folds for each of 1 candidates, totalling 3 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done   3 out of   3 | elapsed:    5.5s finished


Best score:  0.6761613536539723
Best parameters set: 
 {'class_weight': {0: 1, 1: 1}, 'n_estimators': 50}


In [ ]:
names = list(X_train)
feature_importances = pd.DataFrame(grid_rf.best_estimator_.feature_importances_,
                                   index = X_train.columns,
                                    columns=['importance']).sort_values('importance')
print("Features sorted by their score: Top 10")
feature_importances.tail(10)

Features sorted by their score: Top 10


importance
fico_score            0.051514
open_acc              0.052473
loan_amnt             0.061908
int_rate              0.063648
mo_sin_old_il_acct    0.063762
annual_inc            0.064356
revol_util            0.064444
revol_bal             0.065543
earliest_cr_line      0.065823
dti                   0.070515